src/main/scala/firrtl/Compiler.scala

```diff
@@ -26,60 +29,104 @@ object RenameMap {
 
-  /** Get renames of a [[CircuitName]]
-    * @note A [[CircuitName]] can only be renamed to a single [[CircuitName]]
-    */
-  def get(key: CircuitName): Option[CircuitName] = underlying.get(key).map {
-    case Seq(c: CircuitName) => c
-    case other => error(s"Unsupported Circuit rename to $other!")
-  }
-  /** Get renames of a [[ModuleName]]
-    * @note A [[ModuleName]] can only be renamed to one-or-more [[ModuleName]]s
-    */
-  def get(key: ModuleName): Option[Seq[ModuleName]] = {
-    def nestedRename(m: ModuleName): Option[Seq[ModuleName]] =
-      this.get(m.circuit).map(cname => Seq(ModuleName(m.name, cname)))
-    underlying.get(key) match {
-      case Some(names) => Some(names.flatMap {
-        case m: ModuleName =>
-          nestedRename(m).getOrElse(Seq(m))
-        case other => error(s"Unsupported Module rename of $key to $other")
-      })
-      case None => nestedRename(key)
-    }
-  }
-  /** Get renames of a [[ComponentName]]
-    * @note A [[ComponentName]] can only be renamed to one-or-more [[ComponentName]]s
-    */
-  def get(key: ComponentName): Option[Seq[ComponentName]] = {
-    def nestedRename(c: ComponentName): Option[Seq[ComponentName]] =
-      this.get(c.module).map { modules =>
-        modules.map(mname => ComponentName(c.name, mname))
+  def apply(t: Target): Seq[Target] = get(t).getOrElse(Seq(t))
+
+  private def recursiveGet(set: mutable.HashSet[Target], cache: mutable.HashMap[Target, Seq[Target]], errors: mutable.ArrayBuffer[String])(key: Target): Seq[Target] = {
+    if(cache.contains(key)) {
+      cache(key)
+    } else {
+      val remapped = underlying.getOrElse(key, Seq(key))
+      val isCircuitNames = (key +: remapped).forall(_.isCircuitName)
+      val isModuleNames = (key +: remapped).forall(_.isModuleName)
+      val isReferences = (key +: remapped).forall(_.isReference)
+
+      val checked = if(!(isCircuitNames || isModuleNames || isReferences)) {
+        if(key.isReference && remapped.forall(r => r.isModuleName || r.isReference)) {
+          // Ok to map from references to modules
+        } else {
+          errors += s"Cannot rename from $key to $remapped!"
+        }
+        Seq(key)
+      } else if(key.isCircuitName && remapped.size > 1) {
+        errors += s"Cannot rename from $key to $remapped!"
+        Seq(key)
+      } else remapped
+
+      if(set.contains(key) && !key.isCircuitName) {
+        throwInternalError(s"Bad rename: $key is in $set")
       }
-    underlying.get(key) match {
-      case Some(names) => Some(names.flatMap {
-        case c: ComponentName =>
-          nestedRename(c).getOrElse(Seq(c))
-        case other => error(s"Unsupported Component rename of $key to $other")
-      })
-      case None => nestedRename(key)
+      //println("  " * set.size + key.serialize)
+
+      set += key
+      val getter = recursiveGet(set, cache, errors)(_)
+      // First, check whole key
+      // If it matches, then
+      val ret = checked.flatMap {
```

Comment:
Can you walk me through this? I get most of the case classes, but some get confusing.